src/receivers/HTTPModuleFunctions.ts

```diff
@@ -0,0 +1,264 @@
+    if (contentType === 'application/x-www-form-urlencoded') {
+      const parsedQs = qsParse(bodyAsString);
+      const { payload } = parsedQs;
+      if (typeof payload === 'string') {
+        return JSON.parse(payload);
+      }
+      return parsedQs;
+    }
+    return JSON.parse(bodyAsString);
+  }
+
+  public static async parseAndVerifyHTTPRequest(
+    options: RequestVerificationOptions,
+    req: IncomingMessage,
+    _res?: ServerResponse,
+  ): Promise<BufferedIncomingMessage> {
+    const { signingSecret } = options;
+
+    // Consume the readable stream (or use the previously consumed readable stream)
+    const bufferedReq = await HTTPModuleFunctions.bufferIncomingMessage(req);
+
+    if (options.enabled !== undefined && !options.enabled) {
+      // As the validation is disabled, immediately return the bufferred reuest
+      return bufferedReq;
+    }
+
+    // Find the relevant request headers
+    const signature = HTTPModuleFunctions.getHeader(req, 'x-slack-signature');
+    const requestTimestampSec = Number(HTTPModuleFunctions.getHeader(req, 'x-slack-request-timestamp'));
+    if (Number.isNaN(requestTimestampSec)) {
+      throw new Error(
+        `${verifyErrorPrefix}: header x-slack-request-timestamp did not have the expected type (${requestTimestampSec})`,
+      );
+    }
+
+    // Calculate time-dependent values
+    const nowMsFn = options.nowMilliseconds ?? (() => Date.now());
+    const nowMs = nowMsFn();
+    const fiveMinutesAgoSec = Math.floor(nowMs / 1000) - 60 * 5;
+
+    // Enforce verification rules
+
+    // Rule 1: Check staleness
+    if (requestTimestampSec < fiveMinutesAgoSec) {
+      throw new Error(`${verifyErrorPrefix}: stale`);
+    }
+
+    // Rule 2: Check signature
+    // Separate parts of signature
+    const [signatureVersion, signatureHash] = signature.split('=');
+    // Only handle known versions
+    if (signatureVersion !== 'v0') {
+      throw new Error(`${verifyErrorPrefix}: unknown signature version`);
+    }
+    // Compute our own signature hash
+    const hmac = createHmac('sha256', signingSecret);
+    hmac.update(`${signatureVersion}:${requestTimestampSec}:${bufferedReq.rawBody.toString()}`);
+    const ourSignatureHash = hmac.digest('hex');
+    if (!tsscmp(signatureHash, ourSignatureHash)) {
+      throw new Error(`${verifyErrorPrefix}: signature mismatch`);
+    }
+
+    // Checks have passed! Return the value that has a side effect (the buffered request)
+    return bufferedReq;
+  }
+
+  public static isBufferedIncomingMessage(req: IncomingMessage): req is BufferedIncomingMessage {
+    return Buffer.isBuffer((req as BufferedIncomingMessage).rawBody);
+  }
+
+  public static getHeader(req: IncomingMessage, header: string): string {
+    const value = req.headers[header];
+    if (value === undefined || Array.isArray(value)) {
+      throw new Error(`${verifyErrorPrefix}: header ${header} did not have the expected type (${value})`);
```

Comment:
inside of this conditional we know that `value`  will be either `undefined` or `array` which is not the expected type to insert into the formatted string. 

I think that this message also is not entirely clear about the error as is. 

What would you think about something like `... the expected type (received ${typeof value}, expected string)`

Reply:
Sounds good 👍 I would like to avoid changing these functions a lot in this PR but error message improvement would be safe enough.